xarray/testing/strategies.py

```diff
@@ -0,0 +1,761 @@
+        # no data -> generate data to match dims
+        dims = draw(dims)
+        if isinstance(dims, List):
+            dim_names = dims
+            valid_shapes = npst.array_shapes(min_dims=len(dims), max_dims=len(dims))
+            data = draw(np_arrays(shape=draw(valid_shapes)))
+            dim_sizes = {n: l for n, l in zip(dims, data.shape)}
+            coords = draw(coordinate_variables(dim_sizes=dim_sizes))
+
+        else:
+            # should be a mapping of form {dim_names: lengths}
+            dim_names, shape = list(dims.keys()), tuple(dims.values())
+            data = draw(np_arrays(shape=shape))
+            coords = draw(coordinate_variables(dim_sizes=dims))
+
+    elif data is not None and dims is not None:
+        # both data and dims provided -> check drawn examples are compatible
+        dims = draw(dims)
+        if isinstance(dims, List):
+            dim_names = dims
+            data = draw(data)
+            assume(data.ndim == len(dims))
+            dim_sizes = {n: l for n, l in zip(dims, data.shape)}
+        else:
+            # should be a mapping of form {dim_names: lengths}
+            data = draw(data)
+            dim_sizes = dims
+            dim_names, shape = list(dims.keys()), tuple(dims.values())
+            assume(data.shape == shape)
+
+        coords = draw(coordinate_variables(dim_sizes=dim_sizes))
+
+    else:
+        # nothing provided, so generate everything consistently by drawing dims to match data, and coords to match both
+        data = draw(np_arrays())
+        dim_names = draw(dimension_names(min_ndims=data.ndim, max_ndims=data.ndim))
+        dim_sizes = {n: l for n, l in zip(dim_names, data.shape)}
+        coords = draw(coordinate_variables(dim_sizes=dim_sizes))
+
+    return xr.DataArray(
+        data=convert(data),
+        coords=coords,
+        name=name,
+        dims=dim_names,
+        attrs=attrs,
+    )
+
+
+@st.composite
+def data_variables(
+    draw: st.DrawFn,
+    dim_sizes: Mapping[str, int],
+    allowed_names: st.SearchStrategy[str] = None,
+) -> st.SearchStrategy[Mapping[str, xr.Variable]]:
+    """
+    Generates dicts of alignable Variable objects for use as Dataset data variables.
+
+    Requires the hypothesis package to be installed.
+
+    Parameters
+    ----------
+    dim_sizes: Mapping of str to int
+        Sizes of dimensions to use for variables.
+    allowed_names: Strategy generating strings
+        Allowed names for data variables. Needed to avoid conflict with names of coordinate variables & dimensions.
+    """
+    # TODO these shouldn't have the same name as any dimensions or any coordinates...
+    vars = draw(_alignable_variables(dim_sizes=dim_sizes))
+    dim_names = list(dim_sizes.keys())
+
+    # can't have same name as a dimension
+    # TODO this is also used in coordinate_variables so refactor it out into separate function
+    valid_var_names = allowed_names.filter(lambda n: n not in dim_names)
```

Comment:
It *is* possible to create a strategy which never generates repeated names, within the scope of a single example *and `Dataset`*.  It's not the prettiest thing ever, but it has the huge advantage of letting you generate names (etc) on-demand rather than as a list-up-front, which is much better for performance in mutation and shrinking.

```python
@st.composite()
def unique_names(draw, key, strat):
    cache = draw(st.shared(st.builds(defaultdict, set), key="_xarray_name_caches_"))[key]
    name = draw(strat.filter(lambda n: n not in cache))
    cache.add(name)
    return name
```

Reply:
So if I call this repeatedly from multiple strategies, as part of generating one example `Dataset`, will I still get unique names? I'm thinking about `datasets` calling `dimension_names`, then `data_variables`, then `coordinate_variables`, all of which themselves call `unique_names`. The resulting names all need to be mutually unique for a single example `Dataset`.

Reply:
Looking over the calls again, you can get away without any of this just by adding a names parameter to each of these, and supplying an appropriately filtered strategy each time.  Complementary design to "have _alignable_variables return a dict", too.

Reply:
I've refactored to pass an optional names kwarg into each strategy, which seems to work!